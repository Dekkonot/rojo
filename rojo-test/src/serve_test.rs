use std::fs;

use insta::assert_yaml_snapshot;

use crate::serve_util::{intern_read_response, run_serve_test};

#[test]
fn empty() {
    run_serve_test("empty", |session, mut redactions| {
        let info = session.get_api_rojo().unwrap();

        let root_id = info.root_instance_id;

        let info = redactions.redacted_yaml(info);
        assert_yaml_snapshot!(info);

        let read_result = session.get_api_read(root_id).unwrap();
        intern_read_response(&mut redactions, &read_result, root_id);
        let read_result = redactions.redacted_yaml(read_result);
        assert_yaml_snapshot!(read_result);
    });
}

#[test]
fn scripts() {
    run_serve_test("scripts", |session, mut redactions| {
        let info = session.get_api_rojo().unwrap();
        let root_id = info.root_instance_id;

        let info = redactions.redacted_yaml(info);
        assert_yaml_snapshot!(info);

        let read_result = session.get_api_read(root_id).unwrap();
        intern_read_response(&mut redactions, &read_result, root_id);
        let read_result = redactions.redacted_yaml(read_result);
        assert_yaml_snapshot!(read_result);

        fs::write(session.path().join("foo.lua"), "Updated foo!").unwrap();

        let subscribe_result = session.get_api_subscribe(0).unwrap();
        let subscribe_result = redactions.redacted_yaml(subscribe_result);
        assert_yaml_snapshot!(subscribe_result);

        let read_result = session.get_api_read(root_id).unwrap();
        let read_result = redactions.redacted_yaml(read_result);
        assert_yaml_snapshot!(read_result);
    });
}

#[test]
fn just_txt() {
    run_serve_test("just-txt.txt", |session, mut redactions| {
        let info = session.get_api_rojo().unwrap();

        let root_id = info.root_instance_id;
        let info = redactions.redacted_yaml(info);

        assert_yaml_snapshot!(info);

        let read_result = session.get_api_read(root_id).unwrap();
        redactions.intern_iter(read_result.instances.keys().copied());
        let read_result = redactions.redacted_yaml(read_result);

        assert_yaml_snapshot!(read_result);

        fs::write(session.path(), "Changed content!").unwrap();

        // TODO: Directly served files currently don't trigger changed events!

        // let subscribe_result = session.get_api_subscribe(0).unwrap();
        // let subscribe_result = redactions.redacted_yaml(subscribe_result);

        // assert_yaml_snapshot!(subscribe_result);
    });
}
