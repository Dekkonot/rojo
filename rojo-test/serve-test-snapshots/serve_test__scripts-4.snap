---
source: rojo-test/src/serve_test.rs
expression: read_result
---
instances:
  id-2:
    Id: id-2
    Parent: ~
    Name: scripts
    ClassName: Folder
    Properties: {}
    Children:
      - id-3
      - id-4
    Metadata:
      ignoreUnknownInstances: false
  id-3:
    Id: id-3
    Parent: id-2
    Name: bar
    ClassName: Script
    Properties:
      Source:
        Type: String
        Value: "-- Hello, from bar!"
    Children: []
    Metadata:
      ignoreUnknownInstances: false
  id-4:
    Id: id-4
    Parent: id-2
    Name: foo
    ClassName: ModuleScript
    Properties:
      Source:
        Type: String
        Value: Updated foo!
    Children: []
    Metadata:
      ignoreUnknownInstances: false
messageCursor: 1
sessionId: id-1
